### PR TITLE
NAS-104196 / 12.0 / Nas 104196

### PR DESCRIPTION
**Pull request by dkmullen**
Adds user / group quota editors and tables for datasets

### CODE REVIEW

**Comment by anodos325** [comment]
In testing:
[2020/03/27 12:03:19] (WARNING) PoolDatasetService.set_quota():3328 - data: [{'quota_type': 'USER', 'id': 'bobbery', 'quota_value': 524288000}, {'quota_type': 'USEROBJ', 'id': 'bobbery', 'quota_value': 0}]

I submitted USER quota of 500MiB and USEROBJ quota of 0 (no quota). The UI in this case should only send the USER quota to middleware (i.e. only send if >0).

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+export default {
+  users: {
+    table_title: T('User Quotas'),
+    quota_title: T('Set Quotas'),
+    user_title: T('Apply Quotas to Selected Users'),
+    system_select: {
+      placeholder: T('Select Users Cached by this System'),
+      tooltip: T('This list includes only users cached on this system. The search box \
+  (below) can be used to locate all users, including users connected via Active Directory, etc.')
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      tooltip: T('The list only shows users cached on this system. The search box \
  below can locate all users, including users connected with Active Directory or other services.')
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    search: {
+      placeholder: T('Search For Connected Users'),
+      tooltip: T('Search for all connected users). \
+ Type a user name followed by <i>ENTER</i>. A warning will be displayed \
+ if the name does not match a valid user.')
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      tooltip: T('Search for all connected users. \
 Type a user name and press <i>ENTER</i>. A warning is shown \
 if there are no matches.')
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            },
+    data_quota: {
+      placeholder: T('User Data Quota ') + globalHelptext.human_readable.suggestion_label,
+      tooltip: T('Amount of disk space that can be used by the selected users. \
+ Enter 0 (zero) to remove the quota (i.e., to set it to <i>unlimited</i>).') + globalHelptext.human_readable.suggestion_tooltip
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      tooltip: T('Amount of disk space that can be used by the selected users. \
 Entering <code>0</code> (zero) allows all disk space to be used.') + globalHelptext.human_readable.suggestion_tooltip
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    obj_quota: {
+      placeholder: T('User Object Quota'),
+      tooltip: T('Number of objects that can be owned by each of the selected users. \
+ Entering <code>0</code> (zero) removes the quota (i.e., to set it to <i>unlimited</i>).'),
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
 Entering <code>0</code> (zero) allows unlimited objects.'),
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    filter_dialog: {
+      title_show: T('Show All Users'),
+      message_show: T('Show all available users, even those that have not used any data.'),
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      message_show: T('Show all available users, including those that have not used any data.'),
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      button_show: T('Show'),
+
+      title_filter: T('Filter Users'),
+      message_filter: T('Show only those users who are using data. This is the default view.'),
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      message_filter: T('Shows only the users who are using data. This is the default view.'),
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  },
+
+  groups: {
+    table_title: T('Group Quotas'),
+    quota_title: T('Set Quotas'),
+    group_title: T('Apply Quotas to Selected Groups'),
+    system_select: {
+      placeholder: T('Select Groups Cached by this System'),
+      tooltip: T('This list includes only groups cached on this system. The search box \
+  (below) can be used to locate all groups, including groups connected via Active Directory, etc.')
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      tooltip: T('The list shows only groups cached on this system. The search box \
  below can locate all groups, including groups connected with Active Directory or other services.')
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    search: {
+      placeholder: T('Search For Connected Groups'),
+      tooltip: T('Search for all connected groups). \
+ Type a group name followed by <i>ENTER</i>. A warning will be displayed \
+ if the name does not match a valid group.')
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      tooltip: T('Search for all connected groups. \
 Type a group name and press <i>ENTER</i>. A warning is shown \
 if there are no matches.')

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            },
+    data_quota: {
+      placeholder: T('Group Data Quota ') + globalHelptext.human_readable.suggestion_label,
+      tooltip: T('Amount of disk space that can be used by the selected groups. \
+ Enter 0 (zero) to remove the quota (i.e., to set it to <i>unlimited</i>).') + globalHelptext.human_readable.suggestion_tooltip
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
 Entering <code>0</code> (zero) allows all disk space.') + globalHelptext.human_readable.suggestion_tooltip
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    obj_quota: {
+      placeholder: T('Group Object Quota'),
+      tooltip: T('Number of objects that can be owned by each of the selected groups. \
+ Entering <code>0</code> (zero) removes the quota (i.e., to set it to <i>unlimited</i>).'),
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
 Entering <code>0</code> (zero) allows unlimited objects.'),
```

File: src/app/helptext/storage/volumes/datasets/dataset-quotas.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    filter_dialog: {
+      title_show: T('Show All Groups'),
+      message_show: T('Show all available groups, even those that have not used any data.'),
+      button_show: T('Show'),
+
+      title_filter: T('Filter Groups'),
+      message_filter: T('Show only those groups who are using data. This is the default view.'),
```
Comment (Mrt134) [inline src/app/helptext/storage/volumes/datasets/dataset-quotas.ts]:
```suggestion
      message_filter: T('Shows only the groups that are using data. This is the default view.'),
```

**Pull request merged**